### DISCUSSION TITLE
ReturnDict throws KeyError when part of dataclass and dataclasses.asdict() called

### DISCUSSION THREAD

**Discussion by LiamClarkeNZ** [encode/django-rest-framework#9065] [Potential Issue] [answered]
ReturnDict's [constructor expects the kwarg](https://github.com/encode/django-rest-framework/blob/7a9db57eaf829367601fbaac5f1811d1fcb712fb/rest_framework/utils/serializer_helpers.py#L18) `serializer` to always be provided, and if not an exception is thrown. 

So when [the code in dataclasses.asdict](https://github.com/python/cpython/blob/eff2042fac6dc7d6e2f8aeec5a64a3a44a1a9648/Lib/dataclasses.py#L1320C1-L1323C50) that handles dict elements of a dataclass tries to copy it using the constructor, it fails.

This has also bitten us when using further serializers (Avro etc.) that also use the same approach to copy a subclass of `dict`.

Reproduction test case:
```
from typing import Dict, Any

import dataclasses
from django.test import SimpleTestCase
from rest_framework.utils.serializer_helpers import ReturnDict


@dataclasses.dataclass
class Example:
    some_dict: Dict[Any, Any]


class TestReturnDictInDataClass(SimpleTestCase):

    def test_kwargs_error(self):
        rdict = ReturnDict(serializer=None, a=1, b=2)
        example = Example(some_dict=rdict)
        self.assertRaises(KeyError, lambda: dataclasses.asdict(example))
 ```
 
 **Suggestion**
 
 Given that providing a value of `None` for `serializer` doesn't break the instantiation of ReturnDict, perhaps the `pop` could be replaced by a more forgiving approach?

**Comment by auvipy** [comment, marked as answer]
I would be open to review a PR with a proper failing testcase and a draft solution which don't break any existing right behavior

**Reply by LiamClarkeNZ** [reply]
Cheers, will flick up a draft PR soon :)

**Comment by dfm88** [comment]
Any update on this? thanks

**Reply by LiamClarkeNZ** [reply]
Not from my end sorry, I haven't had the time to do so yet, but will give it a crack soon.

**Comment by jamesliberatii** [comment]
This affects ReturnList as well, see: https://github.com/goauthentik/authentik/issues/9430
